File: app/src/main/java/com/example/schooldiary/ui/adapters/diary/DayHolder.java
package com.example.schooldiary.ui.adapters.diary;

import android.util.Log;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.schooldiary.databinding.ItemDayBinding;
import com.example.schooldiary.model.Day;
import com.example.schooldiary.model.Subject;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;
import com.google.firebase.firestore.Source;

import java.util.List;


public class DayHolder extends RecyclerView.ViewHolder {

    private View itemView;

    public DayHolder(@NonNull View itemView) {
        super(itemView);
        this.itemView = itemView;
    }

    private View getItemView() {
        return itemView;
    }

    public synchronized void bindData(Day day) {

        ItemDayBinding binding = ItemDayBinding.bind(itemView);

        TextView dayName = binding.dayName;
        RecyclerView subjectsRecyclerView = binding.subjectsRecyclerView;

        dayName.setText(new StringBuilder().append(day.getName()).append(", 16 September").toString());
        Log.d("DayName", day.getName());

        Query query = FirebaseFirestore.getInstance()
                .collection("days")
                .document(day.getName())
                .collection("subjects")
                .orderBy("serialNumber");

        query.get(Source.CACHE) // this data don't disappear by clear cache, but by clear data of app
                .addOnCompleteListener(task -> {
                    QuerySnapshot querySnapshot = task.getResult();
                    if (!querySnapshot.isEmpty()) {
                        inflateAdapter(querySnapshot, subjectsRecyclerView);
                    } else {
                        query.get(Source.SERVER)
                                .addOnCompleteListener(newTask -> {
                                    QuerySnapshot documentSnapshot = newTask.getResult();
                                    inflateAdapter(documentSnapshot, subjectsRecyclerView);
                                });
                    }
                });
    }

    private void inflateAdapter(QuerySnapshot querySnapshot, RecyclerView subjectsRecyclerView) {
        List<Subject> subjects = querySnapshot.toObjects(Subject.class);
        Log.d("Query on subjects", subjects.size() + "");


        SubjectAdapter adapter = new SubjectAdapter(subjects);
        subjectsRecyclerView.setAdapter(adapter);
        subjectsRecyclerView.setLayoutManager(new LinearLayoutManager(subjectsRecyclerView.getContext()));
        subjectsRecyclerView.setNestedScrollingEnabled(false); //disable scrolling
    }
}

File: app/src/main/java/com/example/schooldiary/ui/adapters/diary/DayHolder.kt
package com.example.schooldiary.ui.adapters.diary

import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.schooldiary.R
import com.example.schooldiary.databinding.ItemDayBinding
import com.example.schooldiary.model.day.Day
import com.example.schooldiary.model.subject.Subject
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.QuerySnapshot
import com.google.firebase.firestore.Source

class DayHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
    private val possibleDays: Set<String> = setOf("Monday", "Tuesday", "Wednesday",
            "Thursday", "Friday", "Saturday")

    fun bindData(day: Day, listener: SubjectHolder.SubjectClickListener) {
        val binding = ItemDayBinding.bind(itemView)

        if (day.name in possibleDays) {
            binding.name.text = itemView.context.getString(R.string.day_of_the_week, day.name)

            val query = FirebaseFirestore.getInstance()
                    .collection("days")
                    .document(day.name)
                    .collection("subjects")
                    .orderBy("serialNumber")

            query.get(Source.CACHE) // cache don't disappear by clear cache, but by clear app data
                    .addOnCompleteListener { task ->
                        val querySnapshot = task.result
                        if (querySnapshot != null && !querySnapshot.isEmpty) {
                            inflateAdapter(querySnapshot, binding.recyclerView, listener)
                        } else {
                            query.get(Source.SERVER)
                                    .addOnCompleteListener { newTask ->
                                        val documentSnapshot = newTask.result
                                        if (documentSnapshot != null && !documentSnapshot.isEmpty) {
                                            inflateAdapter(documentSnapshot,
                                                    binding.recyclerView,
                                                    listener)
                                        }
                                    }
                        }
                    }
        }

    }

    private fun inflateAdapter(snapshot: QuerySnapshot, recyclerView: RecyclerView,
                               listener: SubjectHolder.SubjectClickListener) {
        val subjects: List<Subject> = snapshot.toObjects(Subject::class.java)
        val adapter = SubjectAdapter(subjects, listener)
        recyclerView.let {
            it.adapter = adapter
            it.layoutManager = LinearLayoutManager(recyclerView.context)
            it.isNestedScrollingEnabled = false
        }
    }
}